qr-backup: init at 1.1.4

acuteaangle opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->
This adds [qr-backup](https://github.com/za3k/qr-backup/), a utility to create paper backups of files using QR codes.

Homepage: https://github.com/za3k/qr-backup
License: `CC0-1.0`

~~Currently blocked by upstream issues:~~ Resolved
- https://github.com/za3k/qr-backup/issues/63
- https://github.com/za3k/qr-backup/issues/60

Closes: gh-343799 ("Package request: qr-backup")
## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

acuteaangle committed:
qr-backup: init at 1.1.4

Add `qr-backup`, a utility to create paper backups of files using QR
codes.

Closes: gh-343799 (Package request: qr-backup)

acuteaangle commented:
@za3k Any idea what might cause _every other_ test to fail consistently?

<details><summary>Output</summary><pre>[...]
correct-restore --instructions both 4s
backup-regression --instructions cover 1s
  command: /nix/store/riyz1x9z6mxdrkf3qxdbg3x9gha271k3-qr-backup-1.1.4/bin/qr-backup --skip-checks - --backup-date 2022-09-22 --instructions cover
  result: 792e4a1704c3d36edc8ce2b96d9e17003a2b23feafb612a5ebd1c68ebbf61a1f != 4757a096d77b5c1cc0774f69577d1365d35d3113115901bbddf63d61411c3ed9
correct-restore --instructions cover 4s
backup-regression --instructions none 1s
  command: /nix/store/riyz1x9z6mxdrkf3qxdbg3x9gha271k3-qr-backup-1.1.4/bin/qr-backup --skip-checks - --backup-date 2022-09-22 --instructions none
  result: c952c9fb7fc3fd4876ab9dc4e67848867e94fcb3ebe57942039b4d2bf00fb8bc != 361e065a0addbe02a11385a8bb9510940e6d09cae6f0482223ca2f08013a1d00
correct-restore --instructions none 3s
backup-regression --note (1) 1s
  command: /nix/store/riyz1x9z6mxdrkf3qxdbg3x9gha271k3-qr-backup-1.1.4/bin/qr-backup --skip-checks - --backup-date 2022-09-22 --note A
  result: fed91611f9dac0d50faa33b0fec76d517f876a0618a08a2b9e58bf04e3a7345c != 49191147dbbe7d9a411567b54f74f86b235f387c8c4a28fe68dbcb86dc310fca
correct-restore --note (1) 3s
backup-regression --note (2) 1s
  command: /nix/store/riyz1x9z6mxdrkf3qxdbg3x9gha271k3-qr-backup-1.1.4/bin/qr-backup --skip-checks - --backup-date 2022-09-22 --note A --note B
  result: 80d1bc7418493dced58e19523698ad8d98a18f48b3bbbdea3618c1de35c02f36 != 1dee193b34d777aa2a20097e3e800f2b4206d73e908168a338baa8011f9b23d6
correct-restore --note (2) 3s
backup-regression --num-copies 1s
  command: /nix/store/riyz1x9z6mxdrkf3qxdbg3x9gha271k3-qr-backup-1.1.4/bin/qr-backup --skip-checks - --backup-date 2022-09-22 --num-copies 3
  result: 64918cb781e99364bbb2c8b510bb0649137f41468d093517d6ceb1cb987674a6 != ba5f2936cfe981ffb5fa074f0893f084747527b00b313f0d86a29260b4f02f07
correct-restore --num-copies 3s
backup-regression --page 1s
  command: /nix/store/riyz1x9z6mxdrkf3qxdbg3x9gha271k3-qr-backup-1.1.4/bin/qr-backup --skip-checks - --backup-date 2022-09-22 --page 100 200
  result: b351bac4ac779fa1f29acd0e7f2ba1ab5c34732ff55ef16e2dded67f635b3fa5 != 551cfb9c092bd3c837f0cbf7fb14c7321dfcf5f0d7cb9d12a79e363004be4ce4
correct-restore --page 1s
backup-regression --qr-version 1s
  command: /nix/store/riyz1x9z6mxdrkf3qxdbg3x9gha271k3-qr-backup-1.1.4/bin/qr-backup --skip-checks - --backup-date 2022-09-22 --qr-version 15
  result: 2c779ae1d17f86e8eca3a588c752fcfdc015b158656e2229df15e2401090cb7d != eebf76bd76f62d6d0f097417ae8fd95cb10963c5dc71e68d88740420899a3325
correct-restore --qr-version 3s
backup-regression --scale 1s
  command: /nix/store/riyz1x9z6mxdrkf3qxdbg3x9gha271k3-qr-backup-1.1.4/bin/qr-backup --skip-checks - --backup-date 2022-09-22 --scale 10
  result: ffb1de2af1d6029f1406725639d473ac0b9bd6fb25dd543bf477d22573aeadf7 != 7c402f001634271bdd04e58b7ebd5bbb71b3d249bc378efa4f5f57bb6d26d8de
correct-restore --scale 3s
backup-regression --no-shuffle 1s
  command: /nix/store/riyz1x9z6mxdrkf3qxdbg3x9gha271k3-qr-backup-1.1.4/bin/qr-backup --skip-checks - --backup-date 2022-09-22 --no-shuffle
  result: 8bdb1e5613fadb295233bea746acdc81cf01dab489eb738ecf1c75cabe8836db != 8ce2b34dc9a57a3fb521bd59ba1f0ec1a7d51f8b67c54fc7ad21eec87454fe0f
correct-restore --no-shuffle 3s
backup-regression 1K zeros 1s
  command: /nix/store/riyz1x9z6mxdrkf3qxdbg3x9gha271k3-qr-backup-1.1.4/bin/qr-backup --skip-checks - --backup-date 2022-09-22
  result: ff17a9733e63fb225df81965d8baa62c3412636a20809baa677c105bb24addc1 != c952b0a40d1fa655b367b274672e0a9b1d56fc034fc690230a0ad842bc4eef53
correct-restore 1K zeros 3s
backup-regression 10K zeros 1s
  command: /nix/store/riyz1x9z6mxdrkf3qxdbg3x9gha271k3-qr-backup-1.1.4/bin/qr-backup --skip-checks - --backup-date 2022-09-22
  result: 720cdb63154857fdd39a6b008c88f6ff329ef42dd9eea026bcf95769e4581432 != 1889f9ad99d9d7df224e3e65283d14602a62b14253bb1adff3c99ae38bc9734b
correct-restore 10K zeros 3s
backup-regression 100K zeros 1s
  command: /nix/store/riyz1x9z6mxdrkf3qxdbg3x9gha271k3-qr-backup-1.1.4/bin/qr-backup --skip-checks - --backup-date 2022-09-22
  result: 475a13f5f0b2cf4f7ab9b2bbabbd0f8d2a93802da26d191237594117e2da1dbb != 1ddfe806cd82e74c5d5f025aae2ce9f38577c9d9d2347337a75fe274e9477e2b
[...]</pre></details>

acuteaangle commented:
Seems like it’s all but one regression test and a couple non-regression tests.
```
correct-restore 100b zeros 3s
backup-regression default options 1s
  command: /nix/store/7sp2q6gq4px1z8ndabybvac3qjy2rcqx-qr-backup-1.1.4/bin/qr-backup --skip-checks - --backup-date 2022-09-22 --dpi 300 --compress --erasure-coding --filename file --instructions page --num-copies 1 --output - --page 500 600 --qr-version 10 --scale 5 --shuffle
  result: e08ce722ad6b6d0079a07672e21e1c25b4fae1c981d2f139003d3de7ffab51f2 != 4d2cd56bc5a890f2a87419ceec58d29cf9b576b14020ebbdabdd75597827add7
correct-restore default options 3s
backup-no-regression --no-compress 1s
correct-restore --no-compress 3s
backup-regression --dpi 1s
  command: /nix/store/7sp2q6gq4px1z8ndabybvac3qjy2rcqx-qr-backup-1.1.4/bin/qr-backup --skip-checks - --backup-date 2022-09-22 --dpi 50
  result: 0a1be4e2a2d147701509c0f43f407b9ddeb0cdf95d262fa8ee76713bcb7112d3 != 22e15a09b519d543c209286636c6faf52c5fc800bab611487d4a47a2a477160b
correct-restore --dpi 4s
failing-command --encrypt 1s
  CRITICAL: Conversion from PDF to TIF failed: -

failing-command --encrypt-print-passphrase 1s
  CRITICAL: Conversion from PDF to TIF failed: -
```

acuteaangle commented:
Well, one of the QR codes is _obviously_ different.
[regression.pdf](https://github.com/user-attachments/files/19450447/regression.pdf)
[regression2.pdf](https://github.com/user-attachments/files/19450449/regression2.pdf)
In this case, looking (nearly?) identical to the first one on the same page.

za3k commented:
To be honest, I've never quite understood the idea that tests should run on every machine that installs something, rather than only for developers or as a troubleshooting step. I have added [a qr-backup issue](https://github.com/za3k/qr-backup/issues/67) to have a version of tests that is expected to pass more reliably, for packagers who feel it is necessary to to do this.

[There is a guide I wrote newly as of yesterday](https://github.com/za3k/qr-backup/blob/master/tests/test.py#L20) which I believe should be printing in the output of the tests, explaining how to debug exactly this problem. If it's not, I'm confused why not. Can you paste the output of the first two sets of tests (100b zeros, default options)?

The error is that yes, the contents of the two QR codes are slightly different. This is fixed in master already (in https://github.com/za3k/qr-backup/commit/8d40ed3e88898e85561105dc755a5595b447adee)

If you would like to work around the issue in v1.1.4 (the next release will include this fix), you can switch Python 3.10 or later -- the problem is that different Python versions give (slightly) different gzip output.

acuteaangle commented:
> To be honest, I've never quite understood the idea that tests should run on every machine that installs something, rather than only for developers or as a troubleshooting step.

Running tests on _install_ is fairly uncommon in NixOS, but running tests _after building from source_ is preferred.
Being a standalone Python script blurs those lines a bit.

I don’t believe the tests would run when a typical user installs the package, assuming they haven’t configured their system to build everything from source.
The tests would only run on the system that actually builds the package (fetch the source, apply any Nix-specific patches, move things to the correct directories).

It would be trivial to not run the tests, but it makes it a lot easier to approve package updates for Nixpkgs if there was testing in CI on the Nicole’s side as well.
(There is a bot which can automatically create an update PR and ping package maintainers after an upstream release; having CI on that (and being able to make use of the tests locally on NixOS) makes it way easier to catch NixOS-specific issues in cases you might not try manually.)

> [There is a guide I wrote newly as of yesterday](https://github.com/za3k/qr-backup/blob/master/tests/test.py#L20) which I believe should be printing in the output of the tests, explaining how to debug exactly this problem. If it's not, I'm confused why not.

That guide did print, I cut it for brevity; it is very useful and I followed it until realizing the regression tests are far too fragile for this—however, once I stopped focusing on getting it working, I realized they’re not actually useful or necessary here.

The regression tests are only useful upstream; they say very little about the correctness of the Nix package.
I think it makes the most sense to disable the regression tests on the Nix side, and only keep the roundtrip tests (as checking that the program still runs and produces sensible output on Nix after an update _is_ important)

The two failing `--encrypt*` tests _do_ seem to be a real problem in the current revision of the Nix package; I’ll need to figure out why that’s happening.

acuteaangle commented:
> Can you paste the output of the first two sets of tests (100b zeros, default options)?

This is what was included in my previous log posted here; are you referring to something different?
```
correct-restore 100b zeros 3s
backup-regression default options 1s
  command: /nix/store/7sp2q6gq4px1z8ndabybvac3qjy2rcqx-qr-backup-1.1.4/bin/qr-backup --skip-checks - --backup-date 2022-09-22 --dpi 300 --compress --erasure-coding --filename file --instructions page --num-copies 1 --output - --page 500 600 --qr-version 10 --scale 5 --shuffle
  result: e08ce722ad6b6d0079a07672e21e1c25b4fae1c981d2f139003d3de7ffab51f2 != 4d2cd56bc5a890f2a87419ceec58d29cf9b576b14020ebbdabdd75597827add7
```

za3k commented:
> > Can you paste the output of the first two sets of tests (100b zeros, default options)?
> 
> This is what was included in my previous log posted here; are you referring to something different?

You can ignore that. I was just trying to figure out why debug instructions weren't being printed (they were).

I have added a `--fast` option to the test runner, which skips reproducibility tests (which I renamed from the less clear "regression" tests). See https://github.com/za3k/qr-backup/commit/e7d028eb4fddaa7f8628c88e1604d5e64f546389. No release is planned immediately -- my advice to use Python 3.10+ for tests/nix stands.

acuteaangle commented:
> I have added a --fast option to the test runner, which skips reproducibility tests (which I renamed from the less clear "regression" tests).

Ah, that would be less janky than [my current solution][1].

[1]: https://github.com/NixOS/nixpkgs/blob/14f9fd3f26a4718e1728bd124a6bcc034e8719fd/pkgs/by-name/qr/qr-backup/disable-fragile-regression-tests.patch

acuteaangle commented:
> No release is planned immediately

No problem :) Little point pushing a release just for a feature in tests

za3k commented:
>    #ghostscript # FIXME - listed upstream but don't know what needs it yet

ghostscript is needed by 'convert' and 'zbarimg' to read PDFs.
In turn, that's needed by qr-backup for restore, or for backups with --no-skip-checks (the default)

The failing line is not printing the exact command it's running (will fix) but the commands the test actually runs are:

> python3 qr-backup --skip-checks - --backup-date 2022-09-22 --encrypt PASSWORD
> python3 qr-backup --restore - --encrypt PASSWORD

za3k commented on pkgs/by-name/qr/qr-backup/disable-fragile-regression-tests.patch:
```diff
@@ -0,0 +1,41 @@
+     '1K zeros, self-check': 'c952b0a40d1fa655b367b274672e0a9b1d56fc034fc690230a0ad842bc4eef53',
+     '1K random, self-check': '0b330dbdabb7f987cff35d262d2e8512506eff5ae010b9c9918251de26f5d633',
+ }
++BLESSED_OUTPUT = dict.fromkeys(BLESSED_OUTPUT, None)
```
This is a good solution and I applaud your coding skills. I would not be able to come up with this patch for most codebases.

acuteaangle replied:
Thank you :3
I have a talent for butchering other peoples code in horrible ways to get something to work :)

***

(Admittedly, there were about 5 `sed` regex attempts trying to find a way to somewhat-durably substitute those hashes before it actually clicked that I was trying to manipulate a Python dictionary.)

The janky part of the patch solution was that it affected `test_assert_full_zeros_same()` as well. A loop with a special case could’ve fixed that, but I figured the 1-line patch was the simple and reliable solution.

acuteaangle replied:
Also, `--fast` is surprisingly fast.

I’ve replaced my janky patch with this for now:
```nix
  patches = [
    # Backported from the upstream master branch so we can `doCheck` on
    # the latest release version without needing to vendor a patch to
    # disable the fragile regression tests
    (fetchpatch {
      name = "0001-Rename-from-regression-to-reproducibility-tests.patch";
      url = "https://github.com/za3k/qr-backup/commit/ec8ab373110de62d2e1f6f47c9a34e9e1a73c571.patch";
      hash = "sha256-93wxmCeH3qxZkDLZa9giMGUmVHdXSitXzet6WnAbQRw=";
    })
    (fetchpatch {
      name = "0002-Add-fast-option-to-tests.patch";
      url = "https://github.com/za3k/qr-backup/commit/e7d028eb4fddaa7f8628c88e1604d5e64f546389.patch";
      hash = "sha256-5Ib+bs+ciqH1z0kvHBD7PrefTTwD6nWk+6pEPqxc9uU=";
    })
  ];
```
(Commented as ‘regression’ still since ‘reproducibility’ may have special meaning to Nix users)

acuteaangle commented:
> > #ghostscript # FIXME - listed upstream but don't know what needs it yet
> 
> ghostscript is needed by 'convert' and 'zbarimg' to read PDFs.
> In turn, that's needed by qr-backup for restore, or for backups with --no-skip-checks (the default)

Hmm, if it’s purely a transitive dependency, it shouldn’t need to be listed in the `qr-backup` package for Nixpkgs.

It seems like [`imagemagick` in Nixpkgs has an _optional_ `ghostscript` dependency, which defaults to false][1].

[`zbar` then depends on `imagemagickBig`][2], but not on `ghostscript` directly.

I haven’t found where `imagemagickBig` is defined yet—it’s possible that variant comes with ghostscript enabled.

[1]: https://github.com/NixOS/nixpkgs/blob/nixos-24.11/pkgs/applications/graphics/ImageMagick/default.nix#L12
[2]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/tools/graphics/zbar/default.nix#L6

acuteaangle commented:
>The failing line is not printing the exact command it's running (will fix) but the commands the test actually runs are:
>
> >python3 qr-backup --skip-checks - --backup-date 2022-09-22 --encrypt PASSWORD
> > python3 qr-backup --restore - --encrypt PASSWORD

Those both work flawlessly when run manually—I’m guessing a dependency must be missing from the test environment for some reason? Weird. Welp; can hopefully just binary search that.

acuteaangle commented:
```console
bash-5.2# /nix/store/z04pv5kb84a3ppb86gx6nq5ighjl76fh-qr-backup-1.1.4/bin/qr-backup --skip-checks <<<'test' --backup-date 2022-09-22 --encrypt PASSWORD
CRITICAL: [gpg stdout] b''
CRITICAL: [gpg stderr] b"gpg: Fatal: can't create directory '/homeless-shelter/.gnupg': No such file or directory\n"
CRITICAL: [gpg exit] 2
bash-5.2#
```
*sigh*. That’ll do it.

`gpg` fails during testing because `$HOME` is set to a nonexistent directory during package builds.

za3k commented:
> > > #ghostscript # FIXME - listed upstream but don't know what needs it yet
> > 
> > 
> > ghostscript is needed by 'convert' and 'zbarimg' to read PDFs.
> > In turn, that's needed by qr-backup for restore, or for backups with --no-skip-checks (the default)
> 
> Hmm, if it’s purely a transitive dependency, it shouldn’t need to be listed in the `qr-backup` package for Nixpkgs.

Yeah, but it's a transitive *optional* dependency, which is non-optional for qr-backup, so it should be listed. 

`imagemagickBig` is exactly what we'd want, sounds perfect.

acuteaangle commented:
> imagemagickBig is exactly what we'd want, sounds perfect.

Already listed :)
`zbar` depends on the `Big` variant, so we’re already pulling it in transitively.
So there’s no waste in using the `Big` variant for the direct dependency as well.

acuteaangle commented:
`versionCheckHook` wasn’t working properly because I was calling the `(pre|post)Check` instead of `(pre|post)InstallCheck` hooks in `installCheckPhase`.

`versionCheckHook` is a sanity check that makes sure the intended package version is found in the output of `--version`. Otherwise, if you forget to update a hash, Nix may silently keep using an old version, assuming it already has the source code with that hash and doesn’t need to fetch it again.

Fixed now

za3k commented:
In qr-backup, someone kindly donated https://github.com/za3k/qr-backup/blob/master/installers/flake.nix and the lock. Can those be safely deleted now?

za3k commented:
Thank you kindly for all your hard work on this!

acuteaangle commented:
> In qr-backup, someone kindly donated https://github.com/za3k/qr-backup/blob/master/installers/flake.nix and the lock. Can those be safely deleted now?

I recommend dropping it once this PR is merged—this PR is unaffected by that, and I don’t know of any reason to keep maintaining the flake upstream.

acuteaangle commented on pkgs/by-name/qr/qr-backup/package.nix:
```diff
@@ -0,0 +1,98 @@
+      --replace-fail '"python3", sys.argv[0]' "'$out/bin/qr-backup'"
+  '';
+
+  installPhase = ''
+    install -D -m 755 qr-backup "$out/bin/qr-backup"
+    mv docs/qr-backup.1{.man,} # remove .man suffix
+    installManPage docs/qr-backup.1
+  '';
```
Ahh, whoops

acuteaangle commented on pkgs/by-name/qr/qr-backup/package.nix:
```diff
@@ -0,0 +1,98 @@
+  doInstallCheck = true;
+  nativeCheckInputs = with pkgs; [
+    versionCheckHook
+    which
+    zbar
+  ];
+  installCheckPhase = ''
+    runHook preInstallCheck
+    substituteInPlace tests/test.py \
+      --replace-fail '"python3", "qr-backup"' "'$out/bin/qr-backup'"
+    export GNUPGHOME="$(mktemp -d)"
+    make test-fast
+    runHook postInstallCheck
+  '';
```
Strange; regardless of whether I set this to `checkPhase` or `installCheckPhase`, it still runs inside `installCheckPhase`. However, `versionCheckPhase` never runs unless `runHook preInstallCheck` is run explicitly.

Changing `installCheckPhase` to `checkPhase` and `runHook {pre,post}InstallCheck` to `runHook {pre,post}Check` causes `versionCheckHook` to never run.

<details>
<summary>Output with <code>checkPhase</code> and <code>runHook {pre,post}Check</code></summary>

```
[...]
wrapping `/nix/store/ips7xv8qlsfx7lhl6slyn1sj20r4mz33-qr-backup-1.1.4/bin/qr-backup'...                                                                                                                          
Executing pythonRemoveTestsDir                                                                                                                                                                                   
Finished executing pythonRemoveTestsDir                                                                                                                                                                          
Running phase: installCheckPhase                                                                                                                                                                                 
python3 tests/test.py --fast                                                                                                                                                                                     
correct-restore --page 2s                                                                                                                                                                                        
[...]                                                                                                                                                                       
correct-restore 50K random 39s                                                                                                                                                                                   
0 failures                                                                                                                                                                                                       
installCheckPhase completed in 52 seconds                                                                                                                                                                        
Running phase: pythonCatchConflictsPhase                                                                                                                                                                         
Running phase: pythonRemoveBinBytecodePhase                                                                                                                                                                      
Running phase: pythonImportsCheckPhase                                                                                                                                                                           
Executing pythonImportsCheckPhase                                                                                                                                                                                
/nix/store/ips7xv8qlsfx7lhl6slyn1sj20r4mz33-qr-backup-1.1.4
```
</details>

<details>
<summary>Output with <code>installCheckPhase</code> and <code>runHook {pre,post}InstallCheck</code></summary>

```
[...]
wrapping `/nix/store/m9l8hysi4h20ch0q1pyihzx5idma1z6g-qr-backup-1.1.4/bin/qr-backup'...
Executing pythonRemoveTestsDir
Finished executing pythonRemoveTestsDir
Running phase: installCheckPhase
Executing versionCheckPhase
Successfully managed to find version 1.1.4 in the output of the command /nix/store/m9l8hysi4h20ch0q1pyihzx5idma1z6g-qr-backup-1.1.4/bin/qr-backup --version
qr-backup v1.1.4
Finished versionCheckPhase
python3 tests/test.py --fast
correct-restore --page 2s
[...]
correct-restore 50K random 39s
0 failures
installCheckPhase completed in 53 seconds
Running phase: pythonCatchConflictsPhase
Running phase: pythonRemoveBinBytecodePhase
Running phase: pythonImportsCheckPhase
Executing pythonImportsCheckPhase
/nix/store/m9l8hysi4h20ch0q1pyihzx5idma1z6g-qr-backup-1.1.4
```
</details>

I have determined that it is `runHook preInstallCheck`. `checkPhase` containing `runHook preInstallCheck` and `runHook postCheck` runs `versionCheckPhase`.

~~I’m not sure what the least-janky solution is here.~~
~~- leave it as `installCheckPhase`, since it’s seemingly equivalent for `buildPythonApplication`?~~
~~- run versionCheckHook manually?~~
~~- run the `preInstallCheck` hook inside `checkPhase`?~~

---

I believe I’ve found the cause:
https://github.com/NixOS/nixpkgs/blob/3fa008da69980bd61324ad9e3a825022f0b736f1/pkgs/development/interpreters/python/python2/mk-python-derivation.nix#L254-L258

If `checkPhase` is defined, it replaces `installCheckPhase`, which means `preInstallCheck` never runs, so `versionCheckHook` never runs.

I think leaving this as `installCheckPhase` makes sense here.

(also weird: changing `nativeCheckInputs` to `nativeInstallCheckInputs` causes an error)

Aleksanaa replied:
I think you've only tried moving `nativeCheckInputs` to `nativeInstallCheckInputs`, not separating them. Anyway, since you are running check against output binary, it still makes sense to put here.

> I believe I’ve found the cause:

That's python2

acuteaangle replied:
> That's python2

Oops

---

I did look into this a little more; the docs seemingly imply `checkPhase` is preferred,

https://github.com/NixOS/nixpkgs/blob/ff07dec4ca7751cb8ffb0483684d67f526fbcea2/doc/languages-frameworks/python.section.md?plain=1#L1244-L1251

However, using `checkPhase` causes `versionCheckHook` to not run. I also tried adding `versionCheckHook` to both `nativeCheckInputs` and `nativeInstallCheckInputs`, but no change.

acuteaangle commented:
7aed00932f52 → d32e2637dc6e:
- add missing `{pre,post}Install` hooks
- change spacing around `{pre,post}InstallCheck` hooks for consistency
- don’t refer to `src.tag` in `meta.changelog`, as `src` may be overridden.

---

I’ve kept `installCheckPhase` for now. Let me know if this needs to be changed to `checkPhase`. Currently, `versionCheckHook` stops functioning if `checkPhase` is set, since `buildPythonApplication` overwrites the default `installCheckPhase` with the contents of `checkPhase` here:

https://github.com/NixOS/nixpkgs/blob/9eff0bb895538ed74c5efd742ba836d7950b1ea3/pkgs/development/interpreters/python/mk-python-derivation.nix#L387-L391

This presumably wipes out the default `preInstallCheck` hook, so that would need to be worked around by either running `preInstallCheck` in `checkPhase` (actually `installCheckPhase` anyways) or running `versionCheckHook` directly.
